File: grupa C/olimpiada/nacionalen/2009/2009-05-05-studentsC/submitsC/Log31004/ship.pas
(*
TASK:SHIP
LANG:PASCAL
*)
program capitan_hook_s_ship;
type vector=array[0..10000] of longint;
var n,i,j,m:longint;
    min,sum,wht:int64;
    p,w:vector;
    gr:array[0..10000] of byte;

procedure group(i:longint);
begin
  while gr[p[i]]=0 do
    begin
      gr[p[i]]:=1;
      i:=p[i];
    end;
end;

function weight(ince:longint;pp,wp:vector):int64;
var wh:int64;
    aux:longint;
begin
  wh:=0;
  while pp[ince]<>ince do
    begin
      wh:=wh+wp[ince]+wp[pp[ince]];
      aux:=wp[ince];
      wp[ince]:=wp[pp[ince]];
      wp[pp[ince]]:=aux;
      aux:=pp[pp[ince]];
      pp[pp[ince]]:=pp[ince];
      pp[ince]:=aux;
   end;
  weight:=wh;
end;

begin
{  assign(input,'ship.in'); reset(input);
  assign(output,'ship.out'); rewrite(output);  }
  readln(n);
  for i:= 1 to n do
    read(w[i-1]);
  for i:= 1 to n do
    read(p[i-1]);
  for i:= 0 to n-1 do
    if (i<>p[i])and(gr[i]<>2) then
      begin
        {cream grupa}
        gr[i]:=1;
        group(i);
        min:=10000000000000000;
        {verifcam pentru fiecare inceput pusibil punctajul si selectam
        maximul
        si de ce dracu' vorbesc la pers 1 pl?}
        for j:= 0 to n-1 do
          if gr[j]=1 then
            begin
              wht:=weight(j,p,w);
              if wht<min then min:=wht;
            end;
        for j:= 0 to n do
          if gr[j]=1 then
          gr[j]:=2;
        sum:=sum+min;
      end;
  writeln(sum);
  {close(output);
  close(input); }
end.

File: grupa C/olimpiada/nacionalen/2009/2009-05-05-studentsC/submitsC/Log31005/ship.cpp
/*
TASK:SHIP
LANG:C++
*/
#include <iostream>
using namespace std;
int main()
{
    int n;
    cin >> n;
    int t[100001], p[100001];
    for(int i=0; i<n; i++) cin >> t[i];
    for(int i=0; i<n; i++) cin >> p[i];
    int k=n, b[100001], j=0;
    for(int i=0; i<n; i++){
            if(p[i]==i){ k--; continue;}
            b[j]=t[i];
            j++;
    }
    sort(b, b+k);
    int sum=0;
    for(int i=0; i<k; i++)
    {
            sum+=b[i];
    }
    int v=k/3;
    if(k%3!=0)
       for(int i=0; i<v; i++){ sum+=2*b[i];}
    else for(int i=0; i<v; i++){ sum+=b[i];}
    cout << sum << endl;
    return 0;
}
